daemon/sailjailclient.c

```diff
@@ -1068,30 +1077,29 @@ sailjailclient_validate_argv(const char *exec, const gchar **app_argv, bool use_
      * such as sailjail, invoker, etc -> make an attempt to skip those
      * by looking for argv[0] for command we are about to launch.
      *
-     * While sandboxing aware apps must use absolute path, legacy apps
-     * usually don't and they must reside in BINDIR.
+     * App may also be defined without absolute path, in which case it
+     * must reside in BINDIR and it must not have any 'wrapper'
+     * executables. Thus check the path of app_argv[0] and compare Exec
+     * line to the binary name.
      */
     const char **tpl_argv = (const char **)exec_argv;
-    const char *alternative = app_argv[0] + sizeof (BINDIR "/") - 1;
-    for( ; *tpl_argv; ++tpl_argv ) {
-        if( !g_strcmp0(*tpl_argv, app_argv[0]) )
-            break;
-        if( use_compatibility && !g_strcmp0(*tpl_argv, alternative) )
-            break;
-    }
-
-    if( !*tpl_argv ) {
-        log_err("Exec line does not contain '%s'", *app_argv);
-        goto EXIT;
-    }
+    if( !path_dirname_eq(app_argv[0], BINDIR) ||
+        g_strcmp0(*tpl_argv, path_basename(app_argv[0])) ) {
+        /* App is not specified without path as the first argument,
+         * => there might be 'wrappers' and we match to full path.
+         */
+        for( ; *tpl_argv; ++tpl_argv ) {
+            if( !g_strcmp0(*tpl_argv, app_argv[0]) )
+                break;
+        }
 
-    if( use_compatibility && strncmp(*tpl_argv, BINDIR "/", sizeof BINDIR) ) {
-        char *temp = (char *)*tpl_argv;
-        *tpl_argv = g_strconcat(BINDIR "/", temp, NULL);
-        g_free(temp);
+        if( !*tpl_argv ) {
+            log_err("Exec line does not contain '%s'", *app_argv);
+            goto EXIT;
+        }
     }
 
-    if( !sailjailclient_match_argv(tpl_argv, app_argv) ) {
+    if( !sailjailclient_match_argv(tpl_argv+1, app_argv+1) ) {
```

Comment:
I need to make some adjustments still. It this wants to check that the first word is not a field code so I have to adjust that to some direction.

Reply:
Removed that "Rule out template starting with a field code" test as redundant.